File: csharp/test/Ice/operations/OnewaysAMI.cs
// Copyright (c) ZeroC, Inc.

namespace Ice.operations;

public static class OnewaysAMI
{
    private static void test(bool b) => global::Test.TestHelper.test(b);

    internal static async Task onewaysAMI(Test.MyInterfacePrx proxy)
    {
        Test.MyInterfacePrx p = Test.MyInterfacePrxHelper.uncheckedCast(proxy.ice_oneway());

        await p.ice_pingAsync();

        try
        {
            _ = p.ice_isAAsync("::Test::MyInterface");
            test(false);
        }
        catch (TwowayOnlyException)
        {
        }

        try
        {
            _ = p.ice_idAsync();
            test(false);
        }
        catch (TwowayOnlyException)
        {
        }

        try
        {
            _ = p.ice_idsAsync();
            test(false);
        }
        catch (TwowayOnlyException)
        {
        }

        await p.opVoidAsync();
        await p.opIdempotentAsync();

        // Calling a ["oneway"] operation on a oneway proxy succeeds.
        await p.opOnewayAsync();

        // Calling a ["oneway"] operation on a twoway proxy throws OnewayOnlyException.
        try
        {
            _ = Test.MyInterfacePrxHelper.uncheckedCast(proxy.ice_twoway()).opOnewayAsync();
            test(false);
        }
        catch (OnewayOnlyException)
        {
        }

        try
        {
            _ = p.opByteAsync(0xff, 0x0f);
            test(false);
        }
        catch (TwowayOnlyException)
        {
        }
    }
}

File: swift/test/Ice/operations/OnewaysAMI.swift
// Copyright (c) ZeroC, Inc.

import Foundation
import Ice
import TestCommon

func onewaysAMI(_ helper: TestHelper, _ proxy: MyClassPrx) async throws {
    let p = uncheckedCast(prx: proxy.ice_oneway(), type: MyClassPrx.self)

    //
    // NOTE: Oneway operations are completed when the request is sent
    //
    try await p.opVoidAsync()

    do {
        let _ = try await p.ice_isAAsync(id: ice_staticId(MyClassPrx.self))
        try helper.test(false)
    } catch is Ice.TwowayOnlyException {}

    do {
        let _ = try await p.ice_idsAsync()
        try helper.test(false)
    } catch is Ice.TwowayOnlyException {}

    try await p.opVoidAsync()

    try await p.opIdempotentAsync()

    do {
        let _ = try await p.opByteAsync(p1: 0xFF, p2: 0x0F)
        try helper.test(false)
    } catch is Ice.TwowayOnlyException {}

}
